core/src/test/java/org/ldaptive/referral/FollowAddReferralHandlerTest.java

```diff
@@ -0,0 +1,293 @@
+        .openPredicate(ldapURL -> true)
+        .build(),
+      Duration.ofSeconds(5));
+
+    final FollowAddReferralHandler handler = new FollowAddReferralHandler(
+      10, url -> new MockConnectionFactory(conn));
+    handler.setRequest(request);
+    handler.setHandle(handle);
+    final AddResponse response = AddResponse.builder()
+      .messageID(1)
+      .resultCode(ResultCode.REFERRAL)
+      .referralURLs("ldap://ds" + count.getAndIncrement() + ".ldaptive.org:389/dc=ldaptive,dc=org??sub?")
+      .build();
+
+    try {
+      handler.apply(response);
+      fail("Should have thrown exception");
+    } catch (Exception e) {
+      assertThat(e).isExactlyInstanceOf(IllegalStateException.class);
+      assertThat(
+        e.getMessage()).startsWith("Referral result handler org.ldaptive.referral.FollowAddReferralHandler");
+    }
+  }
+
+
+  @Test(groups = "handlers")
+  public void applyThrowsLdapException()
+  {
+    final AtomicInteger count = new AtomicInteger(1);
+    final ConnectionConfig config = ConnectionConfig.builder().url("ldap://placeholder.ldaptive.org").build();
+    final MockConnection<AddRequest, AddResponse> conn = new MockConnection<>(config);
+    conn.setOpenPredicate(ldapURL -> true);
+    conn.setAddOperationFunction(req -> new DefaultOperationHandle<>(req, conn, Duration.ofSeconds(5)));
+    conn.setWriteConsumer(h -> {
+      if (count.getAndIncrement() == 2) {
+        h.messageID(1);
+        h.sent();
+        h.result(
+          AddResponse.builder()
+            .messageID(1)
+            .resultCode(ResultCode.REFERRAL)
+            .referralURLs("ldap://ds" + count.getAndIncrement() + ".ldaptive.org:389/dc=ldaptive,dc=org??sub?")
+            .build());
+        // close connection since it will be reused by the next handler invocation
+        conn.close();
+      } else {
+        throw new RuntimeException(new LdapException("Test Exception"));
+      }
+    });
+    final AddRequest request = AddRequest.builder()
+      .dn("ou=test,dc=ldaptive,dc=org")
+      .attributes(LdapAttribute.builder().name("uid").values("1").build())
+      .build();
+    final DefaultOperationHandle<AddRequest, AddResponse> handle = new DefaultOperationHandle<>(
+      request,
+      MockConnection.builder(
+          ConnectionConfig.builder().url("ldap://directory.ldaptive.org").build())
+        .openPredicate(ldapURL -> true)
+        .build(),
+      Duration.ofSeconds(5));
+
+    final FollowAddReferralHandler handler = new FollowAddReferralHandler(
+      10, url -> new MockConnectionFactory(conn));
+    handler.setRequest(request);
+    handler.setHandle(handle);
+    final AddResponse response = AddResponse.builder()
+      .messageID(1)
+      .resultCode(ResultCode.REFERRAL)
+      .referralURLs("ldap://ds" + count.getAndIncrement() + ".ldaptive.org:389/dc=ldaptive,dc=org??sub?")
+      .build();
+
+    assertThat(handler.apply(response)).isEqualTo(
```

Comment:
I'm curious why you chose to treat `LdapException` as a non-failure case. It looks like this simply causes referral handling to short-circuit with no signal.

Reply:
The behavior isn't specified and is implementation specific.
I've chosen to return the original result if `LdapException` is thrown at any point during referral chasing.
I suppose that could be a point of configuration, but you really don't have any good options so that behavior seems as good as any.

Reply:
Noting that I pushed a fix that allows configuration of whether the referral handler should throw in this situation.